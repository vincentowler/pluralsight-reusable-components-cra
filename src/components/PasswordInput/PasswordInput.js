import React, { PropTypes } from 'react';
import ProgressBar from '../ProgressBar/ProgressBar';
import EyeIcon from '../EyeIcon/EyeIcon';

/**
 * Full-featured Password Input with integrated label, quality tips, and show password toggle.
 */
class PasswordInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showPassword: false
    }

    this.toggleShowPassword = this.toggleShowPassword.bind(this);
  }

  toggleShowPassword(event) {
    this.setState((prevState) => {
      return { showPassword: !prevState.showPassword };
    });
    event.preventDefault();
  }

  // Default calculateScore method.
  // Return a number from 0 to 100 that represents password quality.
  calculateScore(containsAlpha, containsNumber, containsSpecialChar) {
    let score = 0;
    const password = this.props.value;
    if (!password) return 0;
    if (containsAlpha) score += 10;
    if (containsNumber) score += 10;
    if (containsSpecialChar) score += 10;
    score += password.length > 7 ? 70 : password.length * 10;
    return score;
  }

  render() {
    const {value, error, name, onChange, placeholder, maxLength, minLength, showVisibilityToggle, showQuality, showTips, ...props} = this.props;
    const {showPassword} = this.state;
    const password = value || '';
    const containsAlpha = password.match(/[a-z]/g);
    const containsNumber = password.match(/\d+/g);
    const containsSpecialChar = password.match(/[^a-zA-Z0-9]+/g);
    const score = this.calculateScore(containsAlpha, containsNumber, containsSpecialChar);

    return (
      <div className="fieldset">
        <div style={{ float: 'left', width: 165 }}>
          <label htmlFor={name}>Password</label>
          <input
            type={showPassword ? 'text' : 'password'}
            placeholder={placeholder || ''}
            onChange={onChange}
            value={password}
            maxLength={maxLength}
            name={name}
            {...props} />
          {
            showVisibilityToggle &&
            <a
              href="#"
              onClick={this.toggleShowPassword}
              style={{marginLeft: 5}}>
              <EyeIcon />
            </a>
          }
          <br />
          {error && <div style={{ color: 'red' }}>{error}</div>}
          {
            showQuality && password.length > 0 && <ProgressBar percent={score} />
          }
        </div>
        <div style={{ float: 'left'}}>
          {
            showTips && password.length > 0 &&
            <ul className="unstyled">
              {!containsAlpha && <li>Add alphabetical character.</li>}
              {!containsNumber && <li>Add number.</li>}
              {!containsSpecialChar && <li>Add special character.</li>}
              {minLength && password.length < minLength && <li>Password must be at least {minLength} characters.</li>}
            </ul>
          }
        </div>
        <div style={{ clear: 'both' }}></div>
      </div>
    );
  }
}

PasswordInput.propTypes = {
  /**
   * Password value
   */
  value: PropTypes.string,
  /**
   * Input name
   */
  name: PropTypes.string.isRequired,
  /**
   * Function called when password input value changes
   */
  onChange: PropTypes.func.isRequired,
  /**
   * Max password length accepted
   */
  maxLength: PropTypes.number.isRequired,
  /**
   * Min password length accepted
   */
  minLength: PropTypes.number,
  /**
   * Placeholder displayed when no password is entered
   */
  placeholder: PropTypes.string,
  /**
   * Set to true to show the toggle for displaying the currently entered password
   */
  showVisibilityToggle: PropTypes.bool,
  /**
   * Set to true to display password quality visually via ProgressBar
   */
  showQuality: PropTypes.bool,
  /**
   * Validation error to display
   */
  error: PropTypes.string,
  /**
   * Set to true to show tips for improving the password
   */
  showTips: PropTypes.bool,
  /**
   * Object that overrides the default theme
   */
  theme: PropTypes.object
};

PasswordInput.default = {
  minLength: 5
}

export default PasswordInput;
