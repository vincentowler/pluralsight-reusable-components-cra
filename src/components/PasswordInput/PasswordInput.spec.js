import React from 'react';
import renderer from 'react-test-renderer';
// Note that this import can be commented out if
// we declare the jestSetup in package.json because
// Enzyme functions are globally available for tests.
import {shallow} from 'enzyme';
import PasswordInput from './PasswordInput';

// Behavioral test
test('toggles input type when show/hide password clicked', () => {
  const wrapper = shallow(<PasswordInput
    htmlId="test"
    name="test"
    value=""
    onChange={() => {}}
    showVisibilityToggle
    />
  );

  // Password input should have a type of password initially.
  expect(wrapper.find({type: 'password'})).toHaveLength(1);

  wrapper.find('a').simulate('click');

  // Password input should have a type of text after clicking toggle.
  expect(wrapper.find({type: 'text'})).toHaveLength(1);
});

test('hides password quality by default', () => {
  const tree = renderer.create(<PasswordInput
    htmlId="test"
    name="test"
    onChange={() => {}}
    value="Uisi38#8iad" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('shows password quality when enabled and a password is entered', () => {
  const tree = renderer.create(<PasswordInput
    htmlId="test"
    name="test"
    onChange={() => {}}
    showQuality
    value="Uisi38#8iad" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('hides password quality when enabled but no password is entered', () => {
  const tree = renderer.create(<PasswordInput
    htmlId="test"
    name="test"
    onChange={() => {}}
    showQuality
    value="" />).toJSON();
  expect(tree).toMatchSnapshot();
});
