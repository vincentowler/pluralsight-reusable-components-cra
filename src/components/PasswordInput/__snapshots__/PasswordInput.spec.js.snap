exports[`test should display tips with bad password 1`] = `
<div>
  <div
    style={
      Object {
        "float": "left",
        "width": 165,
      }
    }>
    <div
      style={
        Object {
          "marginBottom": 16,
        }
      }>
      <label
        htmlFor="test">
        Password
      </label>
      <span
        style={
          Object {
            "color": "red",
          }
        }>
         *
      </span>
      <input
        maxLength={50}
        name="test"
        onChange={[Function]}
        placeholder={undefined}
        style={
          Object {
            "display": "inline-block",
          }
        }
        type="password"
        value="a" />
      <br />
    </div>
  </div>
  <div
    style={
      Object {
        "marginLeft": 170,
      }
    }>
    <ul
      style={
        Object {
          "listStyleType": "none",
          "margin": 0,
        }
      }>
      <li>
        Add number.
      </li>
      <li>
        Add special character.
      </li>
      <li>
        Password must be at least 
        5
         characters.
      </li>
    </ul>
  </div>
  <div
    style={
      Object {
        "clear": "both",
      }
    } />
</div>
`;

exports[`test should hide password quality by default 1`] = `
<div>
  <div
    style={
      Object {
        "float": "left",
        "width": 165,
      }
    }>
    <div
      style={
        Object {
          "marginBottom": 16,
        }
      }>
      <label
        htmlFor="test">
        Password
      </label>
      <span
        style={
          Object {
            "color": "red",
          }
        }>
         *
      </span>
      <input
        maxLength={50}
        name="test"
        onChange={[Function]}
        placeholder={undefined}
        style={
          Object {
            "display": "inline-block",
          }
        }
        type="password"
        value="Uisi38#8iad" />
      <br />
    </div>
  </div>
  <div
    style={
      Object {
        "marginLeft": 170,
      }
    } />
  <div
    style={
      Object {
        "clear": "both",
      }
    } />
</div>
`;

exports[`test should hide password quality when enabled but no password is entered 1`] = `
<div>
  <div
    style={
      Object {
        "float": "left",
        "width": 165,
      }
    }>
    <div
      style={
        Object {
          "marginBottom": 16,
        }
      }>
      <label
        htmlFor="test">
        Password
      </label>
      <span
        style={
          Object {
            "color": "red",
          }
        }>
         *
      </span>
      <input
        maxLength={50}
        name="test"
        onChange={[Function]}
        placeholder={undefined}
        style={
          Object {
            "display": "inline-block",
          }
        }
        type="password"
        value="" />
      <br />
    </div>
  </div>
  <div
    style={
      Object {
        "marginLeft": 170,
      }
    } />
  <div
    style={
      Object {
        "clear": "both",
      }
    } />
</div>
`;

exports[`test should show no tips with good password 1`] = `
<div>
  <div
    style={
      Object {
        "float": "left",
        "width": 165,
      }
    }>
    <div
      style={
        Object {
          "marginBottom": 16,
        }
      }>
      <label
        htmlFor="test">
        Password
      </label>
      <span
        style={
          Object {
            "color": "red",
          }
        }>
         *
      </span>
      <input
        maxLength={50}
        name="test"
        onChange={[Function]}
        placeholder={undefined}
        style={
          Object {
            "display": "inline-block",
          }
        }
        type="password"
        value="Uisi38#8iad" />
      <br />
    </div>
  </div>
  <div
    style={
      Object {
        "marginLeft": 170,
      }
    }>
    <ul
      style={
        Object {
          "listStyleType": "none",
          "margin": 0,
        }
      } />
  </div>
  <div
    style={
      Object {
        "clear": "both",
      }
    } />
</div>
`;

exports[`test should show password quality when enabled and a password is entered 1`] = `
<div>
  <div
    style={
      Object {
        "float": "left",
        "width": 165,
      }
    }>
    <div
      style={
        Object {
          "marginBottom": 16,
        }
      }>
      <label
        htmlFor="test">
        Password
      </label>
      <span
        style={
          Object {
            "color": "red",
          }
        }>
         *
      </span>
      <input
        maxLength={50}
        name="test"
        onChange={[Function]}
        placeholder={undefined}
        style={
          Object {
            "display": "inline-block",
          }
        }
        type="password"
        value="Uisi38#8iad" />
      <br />
      <div
        style={
          Object {
            "border": "solid 1px lightgray",
            "width": 100,
          }
        }>
        <div
          style={
            Object {
              "backgroundColor": "green",
              "height": 5,
              "width": 100,
            }
          } />
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "marginLeft": 170,
      }
    } />
  <div
    style={
      Object {
        "clear": "both",
      }
    } />
</div>
`;
